### PR TITLE
added docker configuration and typeORM connection setup

### PR DESCRIPTION
**Pull request by mmarganski**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      MYSQL_DATABASE: ${TYPEORM_DATABASE}
+      MYSQL_USER: ${TYPEORM_USERNAME}
+      MYSQL_PASSWORD: ${TYPEORM_PASSWORD}
+      networks: host
+
+  adminer:
+    image: adminer:4.8.0
+    restart: always
+    ports:
+      - 8081:8080
```
Comment (jpudysz) [inline docker-compose.yml]: enter, this sign means new line

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+RUN yarn install --production=true
+
+COPY . .
+
+COPY --from=development /usr/src/app/dist ./dist
+
+EXPOSE 3002
+
+CMD ["node", "dist/main"]
```
Comment (jpudysz) [inline Dockerfile]: enter, this sign means new line

File: src/app.service.ts [hunk 3]
```diff
@@ -1,9 +1,37 @@
 import { Injectable } from '@nestjs/common'
 import { Request } from 'express'
+import { createConnection, getRepository } from 'typeorm'
+import { User } from './entity/User'
```
Comment (jpudysz) [inline src/app.service.ts]: this is bad import, add `index.ts`

File: src/app.service.ts [hunk 4]
```diff
@@ -1,9 +1,37 @@
...
 
 @Injectable()
 export class AppService {
     getHello(request: Request): string {
         return `Hello! \n World${JSON.stringify(request.body)}`
     }
 }
+
+export const testDBConnection = async () => {
```
Comment (jpudysz) [inline src/app.service.ts]: no console longs inside PR

File: src/app.service.ts [hunk 5]
```diff
@@ -1,9 +1,37 @@
...
+    console.log("connection test:")
+    await createConnection()
+    console.log("DB connected")
+
+}
+
+export const test = async () => {
```
Comment (jpudysz) [inline src/app.service.ts]: same here, I think that this code shouldn't be included into PR

File: src/app.service.ts [hunk 6]
```diff
@@ -1,9 +1,37 @@
...
+  user.messages = []
+
+  await connection.manager.save(user)
+  const users = await connection.manager.find(User)
+
+  const userRepository = getRepository(User)
+  const user2 = userRepository.create({ userId: '10', userName: 'Bob', avatar: 'bar', isActive: false, messages: [] })
+  await userRepository.save(user2)
+
+}
```
Comment (jpudysz) [inline src/app.service.ts]: nl

File: src/main.ts [hunk 7]
```diff
@@ -1,5 +1,7 @@
 import { NestFactory } from '@nestjs/core'
 import { AppModule } from './app.module'
+import "reflect-metadata"
```
Comment (jpudysz) [inline src/main.ts]: import order

File: src/main.ts [hunk 8]
```diff
@@ -8,3 +10,4 @@ const bootstrap = async () => {
 }
 
 bootstrap()
+test()
```
Comment (jpudysz) [inline src/main.ts]: nl

**Review by jpudysz — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,34 @@
+FROM node:14.17-alpine As development
+
+WORKDIR /usr/src/app
+
+COPY package*.json ./
+COPY yarn.lock ./
+
+RUN yarn install --production=false
```
Comment (WadasGrzegorz) [inline Dockerfile]: you can omit ```--production=false``` in this place

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+COPY . .
+
+EXPOSE 3002
+
+RUN npm run build
```
Comment (WadasGrzegorz) [inline Dockerfile]: use yarn instead

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+WORKDIR /usr/src/app
+
+COPY package*.json ./
+COPY yarn.lock ./
+
+RUN yarn install --production=false
+
+COPY . .
+
+EXPOSE 3002
```
Comment (WadasGrzegorz) [inline Dockerfile]: remove it - development is only for build

File: Dockerfile (hunk 2 above)
Comment (WadasGrzegorz) [inline Dockerfile]: nl

File: docker-compose.yml (hunk 1 above)
Comment (WadasGrzegorz) [inline docker-compose.yml]: nl

File: ormconfig.json [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  ],
+  "migrations": [
+    "dist/migration/*.js"
+  ],
+  "cli": {
+    "entitiesDir": "dist/entity",
+    "migrationsDir": "dist/migration",
+    "subscribersDir": "dist/subscriber"
+  }
+}
```
Comment (WadasGrzegorz) [inline ormconfig.json]: nl

File: src/app.service.ts (hunk 3 above)
Comment (WadasGrzegorz) [inline src/app.service.ts]: entity should have index

File: src/app.service.ts (hunk 6 above)
Comment (WadasGrzegorz) [inline src/app.service.ts]: nl

File: src/app.service.ts [hunk 13]
```diff
@@ -1,9 +1,37 @@
...
+  user.isActive = true
+  user.messages = []
+
+  await connection.manager.save(user)
+  const users = await connection.manager.find(User)
+
+  const userRepository = getRepository(User)
+  const user2 = userRepository.create({ userId: '10', userName: 'Bob', avatar: 'bar', isActive: false, messages: [] })
+  await userRepository.save(user2)
+
```
Comment (WadasGrzegorz) [inline src/app.service.ts]: .

File: src/entity/User.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+  @Column()
+  avatar: string
+
+  @Column()
+  isActive: boolean
+
+  @OneToMany(() => Message, message => message.userId)
+  messages: Message[]
+}
```
Comment (WadasGrzegorz) [inline src/entity/User.ts]: .

File: src/main.ts (hunk 8 above)
Comment (WadasGrzegorz) [inline src/main.ts]: remove it

File: src/main.ts (hunk 7 above)
Comment (WadasGrzegorz) [inline src/main.ts]: ?

File: src/main.ts [hunk 15]
```diff
@@ -1,5 +1,7 @@
 import { NestFactory } from '@nestjs/core'
 import { AppModule } from './app.module'
+import "reflect-metadata"
+import { testDBConnection, test } from './app.service'
```
Comment (WadasGrzegorz) [inline src/main.ts]: .

File: src/entity/User.ts [hunk 16]
```diff
@@ -0,0 +1,21 @@
+import { Column, Entity, OneToMany, PrimaryColumn } from 'typeorm'
+import { Message } from './Message'
+
+@Entity()
+export class User {
+
```
Comment (WadasGrzegorz) [inline src/entity/User.ts]: .

File: src/entity/User.ts [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+  @PrimaryColumn()
```
Comment (WadasGrzegorz) [inline src/entity/User.ts]: You can use `PrimaryGeneratedColumn` instead

File: src/entity/User.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+  userId: string
```
Comment (WadasGrzegorz) [inline src/entity/User.ts]: Id should be number

File: src/entity/Room.ts [hunk 19]
```diff
@@ -0,0 +1,20 @@
+import { Column, Entity, JoinTable, ManyToMany, OneToMany, PrimaryColumn } from 'typeorm'
+import { User } from './User'
+import { Message } from './Message'
+
+@Entity()
+export class Room {
+
```
Comment (WadasGrzegorz) [inline src/entity/Room.ts]: .

File: src/entity/Message.ts [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+  @Column()
+  date: Date
+
+  @ManyToOne(() => User, user => user.messages)
+  userId: string
+
+  @ManyToOne(() => Room, room => room.messages)
+  roomName: string
+
+}
```
Comment (WadasGrzegorz) [inline src/entity/Message.ts]: .

File: src/entity/Message.ts [hunk 21]
```diff
@@ -0,0 +1,26 @@
+import { Column, Entity, ManyToOne, PrimaryGeneratedColumn } from 'typeorm'
+import { User } from './User'
+import { Room } from './Room'
+
+@Entity()
+export class Message {
+
```
Comment (WadasGrzegorz) [inline src/entity/Message.ts]: .

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

File: src/app.controller.ts [hunk 22]
```diff
@@ -1,18 +1,8 @@
 import { Controller, Get, Req, Post } from '@nestjs/common'
 import { Request } from 'express'
-import { AppService } from './app.service'
+// import { AppService } from './app.service'
```
Comment (jpudysz) [inline src/app.controller.ts]: dont push commented code

File: src/entity/Message.ts [hunk 23]
```diff
@@ -0,0 +1,24 @@
+import { Column, Entity, ManyToOne, PrimaryGeneratedColumn } from 'typeorm'
+import { User, Room } from './index'
```
Comment (jpudysz) [inline src/entity/Message.ts]: dont import from index, you can import from file if you are in the same folder

File: src/entity/Room.ts [hunk 24]
```diff
@@ -0,0 +1,19 @@
+import { Column, Entity, JoinTable, ManyToMany, OneToMany, PrimaryColumn } from 'typeorm'
+import { User, Message } from './index'
```
Comment (jpudysz) [inline src/entity/Room.ts]: aa

File: src/entity/User.ts [hunk 25]
```diff
@@ -0,0 +1,21 @@
+import { Column, Entity, OneToMany, PrimaryColumn } from 'typeorm'
+import { Message } from './index'
```
Comment (jpudysz) [inline src/entity/User.ts]: aa

**Review by jpudysz — CHANGES_REQUESTED** [review]

File: package.json [hunk 26]
```diff
@@ -25,11 +25,14 @@
     "@nestjs/core": "8.0.0",
     "@nestjs/platform-express": "8.0.0",
     "@nestjs/platform-socket.io": "8.0.5",
+    "@nestjs/typeorm": "^8.0.2",
```
Comment (jpudysz) [inline package.json]: .

File: src/lib/entities/Message.entity.ts [hunk 27]
```diff
@@ -0,0 +1,24 @@
+import { Column, Entity, ManyToOne, PrimaryGeneratedColumn } from 'typeorm'
+import { RoomEntity } from './'
```
Comment (jpudysz) [inline src/lib/entities/Message.entity.ts]: .

File: src/lib/entities/Room.entity.ts [hunk 28]
```diff
@@ -0,0 +1,15 @@
+import { Column, Entity, OneToMany, PrimaryColumn } from 'typeorm'
+import { MessageEntity } from './'
```
Comment (jpudysz) [inline src/lib/entities/Room.entity.ts]: not like that, use `'./MessageEntity'`

File: src/main.ts [hunk 29]
```diff
@@ -1,5 +1,5 @@
 import { NestFactory } from '@nestjs/core'
-import { AppModule } from './app.module'
+import { AppModule } from './modules/app/app.module'
```
Comment (jpudysz) [inline src/main.ts]: bad, nested import, use index.ts

File: src/modules/app/app.module.ts [hunk 30]
```diff
@@ -0,0 +1,22 @@
+import { Module } from '@nestjs/common'
+import { TypeOrmModule } from '@nestjs/typeorm'
+import { AppService } from './app.service'
+import { AppGateway } from './app.gateway'
+import { Connection, getConnectionOptions } from 'typeorm'
```
Comment (jpudysz) [inline src/modules/app/app.module.ts]: import order

File: src/modules/app/app.module.ts [hunk 31]
```diff
@@ -0,0 +1,22 @@
...
+import { MessageEntity, RoomEntity, UserEntity } from '../../lib/entities'
```
Comment (jpudysz) [inline src/modules/app/app.module.ts]: bad import

File: src/modules/app/app.service.ts [hunk 32]
```diff
@@ -0,0 +1,17 @@
+import { Connection, Repository } from 'typeorm'
+import { UserEntity, RoomEntity, MessageEntity } from 'lib/entities'
+import { InjectRepository } from '@nestjs/typeorm'
```
Comment (jpudysz) [inline src/modules/app/app.service.ts:1]: import order

**Review by jpudysz — CHANGES_REQUESTED** [review]

File: src/lib/entities/index.ts [hunk 33]
```diff
@@ -0,0 +1,4 @@
+export { UserEntity } from './User.entity'
+export { RoomEntity } from './Room.entity'
+export { MessageEntity } from './Message.entity'
+export { UserRoomEntity } from './UserRoom.entity'
```
Comment (jpudysz) [inline src/lib/entities/index.ts]: nl

File: src/modules/app/app.module.ts [hunk 34]
```diff
@@ -0,0 +1,22 @@
+import { Module } from '@nestjs/common'
+import { TypeOrmModule } from '@nestjs/typeorm'
+import { Connection, getConnectionOptions } from 'typeorm'
+import { AppService } from './app.service'
+import { AppGateway } from './app.gateway'
+import { MessageEntity, RoomEntity, UserEntity } from 'lib/entities'
```
Comment (jpudysz) [inline src/modules/app/app.module.ts]: import order

**Review by jpudysz — CHANGES_REQUESTED** [review]

File: ormconfig.json [hunk 35]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (WadasGrzegorz) [inline ormconfig.json]: this file shouldn't be in GitHub - what if you had there production config ?

File: src/lib/entities/Message.entity.ts [hunk 36]
```diff
@@ -0,0 +1,24 @@
+import { Column, Entity, ManyToOne, PrimaryGeneratedColumn } from 'typeorm'
+import { RoomEntity } from './Room.entity'
+
+@Entity({name: 'Message'})
+export class MessageEntity {
+
```
Comment (WadasGrzegorz) [inline src/lib/entities/Message.entity.ts]: rm empty line

File: src/lib/entities/Message.entity.ts [hunk 37]
```diff
@@ -0,0 +1,24 @@
+import { Column, Entity, ManyToOne, PrimaryGeneratedColumn } from 'typeorm'
+import { RoomEntity } from './Room.entity'
```
Comment (WadasGrzegorz) [inline src/lib/entities/Message.entity.ts]: file name should starts with lowercase

File: src/lib/entities/Message.entity.ts [hunk 38]
```diff
@@ -0,0 +1,24 @@
...
+
+@Entity({name: 'Message'})
```
Comment (WadasGrzegorz) [inline src/lib/entities/Message.entity.ts]: table name should start with lowercase and spaces around

File: src/lib/entities/Room.entity.ts [hunk 39]
```diff
@@ -0,0 +1,15 @@
+import { Column, Entity, OneToMany, PrimaryColumn } from 'typeorm'
+import { MessageEntity } from './Message.entity'
+
+@Entity({name: 'Room'})
```
Comment (WadasGrzegorz) [inline src/lib/entities/Room.entity.ts]: whole file aa

File: src/lib/entities/User.entity.ts [hunk 40]
```diff
@@ -0,0 +1,17 @@
+import { Column, Entity, PrimaryColumn } from 'typeorm'
```
Comment (WadasGrzegorz) [inline src/lib/entities/User.entity.ts]: aa

File: src/lib/entities/UserRoom.entity.ts [hunk 41]
```diff
@@ -0,0 +1,21 @@
+import { Column, Entity, PrimaryGeneratedColumn, Unique } from 'typeorm'
```
Comment (WadasGrzegorz) [inline src/lib/entities/UserRoom.entity.ts]: aa

File: src/lib/entities/index.ts (hunk 33 above)
Comment (WadasGrzegorz) [inline src/lib/entities/index.ts]: nl

File: src/modules/app/app.module.ts [hunk 42]
```diff
@@ -0,0 +1,22 @@
...
+                Object.assign(await getConnectionOptions(), {
+                    autoLoadEntities: true,
+                })
+        }),
+        TypeOrmModule.forFeature([UserEntity, RoomEntity, MessageEntity])],
+    controllers: [],
+    providers: [AppService, AppGateway]
+})
+export class AppModule {
+    constructor(private readonly connection: Connection) {}
```
Comment (WadasGrzegorz) [inline src/modules/app/app.module.ts]: ?

File: src/modules/app/app.service.ts [hunk 43]
```diff
@@ -0,0 +1,113 @@
...
+import { Injectable } from '@nestjs/common'
+import { Connection, Repository } from 'typeorm'
+import { UserEntity, RoomEntity, MessageEntity } from 'lib/entities'
+
+@Injectable()
+export class AppService{
+    constructor(
+        private readonly connection: Connection,
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: rm

File: src/modules/app/app.service.ts [hunk 44]
```diff
@@ -0,0 +1,113 @@
...
+    async getUsersList() {
+        return this.userRepository.find({ select: ['socketId', 'userName', 'avatar', 'isActive'] })
+    }
+
+    async getRoomNames() {
+        return this.roomRepository.find({ select: ['roomName'] })
+    }
+
+    async getPublicRooms() {
+        return this.roomRepository.find({select: ['roomName'], where: {isPrivate: false}})
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: spaces

File: src/modules/app/app.service.ts [hunk 45]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    async getUsersByRoomName (roomName: string) {
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: just ```
         return this.roomRepository.find({ where: { roomName } })```

File: src/modules/app/app.service.ts [hunk 46]
```diff
@@ -0,0 +1,113 @@
...
+    async getMessagesByRoomName(roomName: string) {
+        return this.messageRepository.find({
+            select: ['messageContent', 'isImage', 'date', 'socketId'],
+            where: {roomName}
+        })
+    }
+
+    async getUserBySocketId(socketId: string) {
+        const userRepository = this.userRepository
+        const [user] = await userRepository.find({where: { socketId}})
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: aa and you can use findOne

File: src/modules/app/app.service.ts [hunk 47]
```diff
@@ -0,0 +1,113 @@
...
+
+        return user
+    }
+
+    async getRoomByName(roomName: string) {
+        const roomRepository = this.roomRepository
+        const [room] = await roomRepository.find({where: {roomName}})
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: aa

File: src/modules/app/app.service.ts [hunk 48]
```diff
@@ -0,0 +1,113 @@
...
+
+        return room
+    }
+
+    async createUser(socketId: string, userName: string, avatar: string) {
+        const userRepository = this.userRepository
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]:
```  
 return this.userRepository.save({
             socketId,
             userName,
             avatar,
             isActive: true
         })
```

File: src/modules/app/app.service.ts [hunk 49]
```diff
@@ -0,0 +1,113 @@
...
+            userName,
+            avatar,
+            isActive: true
+        })
+
+        return userRepository.save(user)
+    }
+
+    async createRoom (roomName: string, isPrivate: boolean) {
+        const roomRepository = this.roomRepository
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: aa

File: src/modules/app/app.service.ts [hunk 50]
```diff
@@ -0,0 +1,113 @@
...
+        const room = roomRepository.create({
+            roomName,
+            isPrivate
+        })
+
+        return roomRepository.save(room)
+    }
+
+    async createMessage (messageContent: string, isImage: boolean, date: Date, userId: string, roomName: string) {
+        const messageRepository = this.messageRepository
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: aa

File: src/modules/app/app.service.ts [hunk 51]
```diff
@@ -0,0 +1,113 @@
...
+        return userRepository.save(user)
+    }
+
+    async joinRoom (socketId: string, roomName: string) {
+        const userRepository = this.userRepository
+        const roomRepository = this.roomRepository
+        const [user] = await userRepository.find({where: {socketId}})
+        const [room] = await roomRepository.find({where: {roomName}})
+
+        // room.users = [...room.users, user]
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: .

File: src/modules/app/app.service.ts [hunk 52]
```diff
@@ -0,0 +1,113 @@
...
+
+        user.isActive = false
+
+        return userRepository.save(user)
+    }
+
+    async joinRoom (socketId: string, roomName: string) {
+        const userRepository = this.userRepository
+        const roomRepository = this.roomRepository
+        const [user] = await userRepository.find({where: {socketId}})
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: aa

File: src/modules/app/app.service.ts [hunk 53]
```diff
@@ -0,0 +1,113 @@
...
+            roomName
+        })
+
+        return messageRepository.save(message)
+    }
+
+    async deactivateUser (socketId: string) {
+        const userRepository = this.userRepository
+        const [user] = await userRepository.find({where: { socketId}})
+
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: in this place you can use update where instead two queries you make only one

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

File: src/modules/app/app.service.ts [hunk 54]
```diff
@@ -0,0 +1,79 @@
...
+    constructor(
+        @InjectRepository(UserEntity)
+        private readonly userRepository: Repository<UserEntity>,
+        @InjectRepository(RoomEntity)
+        private readonly roomRepository: Repository<RoomEntity>,
+        @InjectRepository(MessageEntity)
+        private readonly messageRepository: Repository<MessageEntity>,
+    ) {}
+
+    async getUsersList() {
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: you don't need `async` if you not using `await`

File: src/modules/app/app.service.ts [hunk 55]
```diff
@@ -0,0 +1,79 @@
...
+        return this.userRepository.find({ select: ['socketId', 'userName', 'avatar', 'isActive'] })
+    }
+
+    async getRoomNames() {
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: aa

File: src/modules/app/app.service.ts [hunk 56]
```diff
@@ -0,0 +1,79 @@
...
+    async getMessagesByRoomName(roomName: string) {
+        return this.messageRepository.find({
+            select: ['messageContent', 'isImage', 'date', 'socketId'],
+            where: { roomName }
+        })
+    }
+
+    async getUserBySocketId(socketId: string) {
+        return this.userRepository.findOne({ where: { socketId } })
+
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: . :)

File: src/modules/app/app.service.ts [hunk 57]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    async getRoomByName(roomName: string) {
+        return this.roomRepository.findOne({ where: { roomName } })
+
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: aa

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

File: src/lib/entities/index.ts [hunk 58]
```diff
@@ -0,0 +1,4 @@
+export { UserEntity } from './user.entity'
+export { RoomEntity } from './room.entity'
+export { MessageEntity } from './message.entity'
+export { UserRoomEntity } from './userRoom.entity'
```
Comment (WadasGrzegorz) [inline src/lib/entities/index.ts]: it is work ?

File: src/modules/app/app.service.ts [hunk 59]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    getUserBySocketId(socketId: string) {
+        return this.userRepository.findOne({ where: { socketId } })
+
+    }
+
+    getRoomByName(roomName: string) {
+        return this.roomRepository.findOne({ where: { roomName } })
+
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: remove empty line after return

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

**Review by WadasGrzegorz — APPROVED** [review]

**Pull request merged**